src/gc.c

```diff
@@ -271,8 +271,12 @@ HL_PRIM void hl_add_root( void *r ) {
 	if( gc_roots_count == gc_roots_max ) {
 		int nroots = gc_roots_max ? (gc_roots_max << 1) : 16;
 		void ***roots = (void***)malloc(sizeof(void*)*nroots);
-		memcpy(roots,gc_roots,sizeof(void*)*gc_roots_count);
-		free(gc_roots);
+
+		if(gc_roots) {

```

Comment:
I wonder: is this assumption ever violated? Can we use an `assume` intrinsic or similar?

Reply:
Well, when hashlink starts gc_roots == NULL.

```
frame #5: 0x00007ffff7e0abc7 libhl.so.1`hl_add_root(r=0x00007ffff7f1c1a0) at gc.c:274:3
   271          if( gc_roots_count == gc_roots_max ) {
   272                  int nroots = gc_roots_max ? (gc_roots_max << 1) : 16;
   273                  void ***roots = (void***)malloc(sizeof(void*)*nroots);
-> 274                  memcpy(roots,gc_roots,sizeof(void*)*gc_roots_count);
   275                  free(gc_roots);
   276                  gc_roots = roots;
   277                  gc_roots_max = nroots;
(lldb) v
(void *) r = 0x00007ffff7f1c1a0
(int) nroots = 16
(void ***) roots = 0x0000000001194c60
(lldb) print gc_roots
(void ***) 0x0000000000000000
```

So, we can add that check or initialize it before calling hl_add_root() somewhere else.